compiler/src/dotty/tools/dotc/transform/Splicing.scala

```diff
@@ -0,0 +1,394 @@
+          Some((fn, ArgsClause(args) :: Nil))
+        case GenericApply(CapturedApplication(fn, argss), args) =>
+          Some((fn, argss :+ ArgsClause(args)))
+        case _ =>
+          None
+    }
+
+    private def containsCapturedType(tpe: Type)(using Context): Boolean =
+      tpe match
+        case tpe @ TypeRef(prefix, _) => isCaptured(tpe.symbol) || containsCapturedType(prefix)
+        case tpe @ TermRef(prefix, _) => isCaptured(tpe.symbol) || containsCapturedType(prefix)
+        case AppliedType(tycon, args) => containsCapturedType(tycon) || args.exists(containsCapturedType)
+        case _ => false
+
+    /** Transform an assignment `x = e` with a captured `x` to
+     * `${ Assign(x$1.asTerm, '{e}.asTerm).asExpr.asInstanceOf[Expr[T]] }`
+     *
+     *  Registers `x` as a captured variable in the hole and creates an `x$1` `Expr` reference to it.
+     */
+    private def transformSplicedAssign(tree: Assign)(using Context): Tree =
+      spliced(tree.tpe) {
+        reflect.asExpr(tree.tpe)(
+          reflect.Assign(
+            reflect.asTerm(capturedTerm(tree.lhs)),
+            reflect.asTerm(quoted(tree.rhs))
+          )
+        )
+      }
+
+    /** Transform an application `f(a1, a2, ...)` with a captured `f` to
+     * `${ Apply(f$1.asTerm, List('{a1$}.asTerm, '{a2$}.asTerm, ...)).asExpr.asInstanceOf[Expr[T]] }`
+     *
+     *  Registers `f` as a captured variable in the hole and creates an `f$1` `Expr` reference to it.
+     *
+     *  It also handles cases with multiple argument clauses using nested `Apply`/`TypeApply`.
+     */
+    private def transformCapturedApplication(tree: Tree, fn: RefTree, argss: List[ArgsClause])(using Context): Tree =
+      spliced(tree.tpe) {
+        def TermList(args: List[Tree]): List[Tree] =
+          args.map(arg => reflect.asTerm(quoted(transform(arg)(using spliceContext))))
+        def TypeTreeList(args: List[Tree]): List[Tree] =
+          args.map(arg => reflect.Inferred(reflect.TypeReprOf(transform(arg)(using spliceContext).tpe)))
+        reflect.asExpr(tree.tpe) {
+          argss.foldLeft[Tree](reflect.asTerm(capturedTerm(fn, defn.AnyType))) { case (acc, clause) =>
+            if clause.isTerm then reflect.Apply(acc, TermList(clause.args))
+            else reflect.TypeApply(acc, TypeTreeList(clause.args))
+          }
+        }
+      }
+
+    private def capturedTerm(tree: Tree)(using Context): Tree =
+      val tpe = tree.tpe.widenTermRefExpr match {
+        case tpw: MethodicType => tpw.toFunctionType(isJava = false)
+        case tpw => tpw
+      }
+      capturedTerm(tree, tpe)
+
+    private def capturedTerm(tree: Tree, tpe: Type)(using Context): Tree =
+      def newBinding = newSymbol(
+        spliceOwner,
+        UniqueName.fresh(tree.symbol.name.toTermName).toTermName,
+        Param,
+        defn.QuotedExprClass.typeRef.appliedTo(tpe),
+      )
+      val (_, bindingSym) = refBindingMap.getOrElseUpdate(tree.symbol, (tree, newBinding))
```

Comment:
I would go with:
```suggestion
      val bindingSym = refBindingMap.getOrElseUpdate(tree.symbol, (tree, newBinding))._2
```
(same below)

Reply:
Done